frame/balances/src/lib.rs

```diff
@@ -626,87 +606,110 @@ impl<T: Trait<I>, I: Instance> Module<T, I> {
-	/// the caller will do this.
-	fn set_account(
+	/// Returns the final free balance, iff the account was previously of total balance zero, known
+	/// as its "endowment".
+	fn post_mutation(
 		who: &T::AccountId,
-		account: &AccountData<T::Balance>,
-		old: &AccountData<T::Balance>,
-	) -> UpdateBalanceOutcome {
-		let total = account.free + account.reserved;
+		new: AccountData<T::Balance>,
+	) -> Option<AccountData<T::Balance>> {
+		let total = new.total();
 		if total < T::ExistentialDeposit::get() {
-			T::DustRemoval::on_unbalanced(NegativeImbalance::new(total));
-			if !old.total().is_zero() {
-				Self::reap_account(who, total);
-				UpdateBalanceOutcome::AccountKilled
-			} else {
-				UpdateBalanceOutcome::StillDead
+			if !total.is_zero() {
+				T::DustRemoval::on_unbalanced(NegativeImbalance::new(total));
+				Self::deposit_event(RawEvent::DustLost(who.clone(), total));
 			}
+			None
 		} else {
-			if old.total().is_zero() {
-				Self::about_to_create_account(who, account.free);
-			}
-			Account::<T, I>::insert(who, account);
-			UpdateBalanceOutcome::Updated
+			Some(new)
 		}
 	}
 
-	/// Register a new account (with existential balance).
+	/// Mutate an account to some new value, or delete it entirely with `None`. Will enforce
+	/// `ExistentialDeposit` law, annulling the account as needed.
+	///
+	/// NOTE: Doesn't do any preparatory work for creating a new account, so should only be used
+	/// when it is known that the account already exists.
 	///
-	/// This just calls appropriate hooks. It doesn't (necessarily) make any state changes.
-	fn about_to_create_account(who: &T::AccountId, balance: T::Balance) {
-		T::OnNewAccount::on_new_account(&who);
-		Self::deposit_event(RawEvent::NewAccount(who.clone(), balance.clone()));
+	/// NOTE: LOW-LEVEL: This will not attempt to maintain total issuance. It is expected that
+	/// the caller will do this.
+	fn mutate_account<R>(
+		who: &T::AccountId,
+		f: impl FnOnce(&mut AccountData<T::Balance>) -> R
+	) -> R {
+		Self::try_mutate_account(who, |a| -> Result<R, Infallible> { Ok(f(a)) })
+			.expect("Error is infallible; qed")
 	}
 
-	/// Unregister an account.
+	/// Mutate an account to some new value, or delete it entirely with `None`. Will enforce
+	/// `ExistentialDeposit` law, annulling the account as needed. This will do nothing if the
+	/// result of `f` is an `Err`.
 	///
-	/// This just removes the nonce and leaves an event.
-	fn reap_account(who: &T::AccountId, dust: T::Balance) {
-		Locks::<T, I>::remove(who);
-		Account::<T, I>::remove(who);
-		T::OnReapAccount::on_reap_account(who);
-		Self::deposit_event(RawEvent::ReapedAccount(who.clone(), dust));
+	/// NOTE: Doesn't do any preparatory work for creating a new account, so should only be used
+	/// when it is known that the account already exists.
+	///
+	/// NOTE: LOW-LEVEL: This will not attempt to maintain total issuance. It is expected that
+	/// the caller will do this.
+	fn try_mutate_account<R, E>(
```

Comment:
I would want to double check all this logic and the interactions with the `AccountStore` impl in system. It looks right, but is not super easy to follow tbh